{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ce1cd3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "── \u001b[1mInstalled datasets\u001b[22m ────────────────────────────────────────────────────────────────────────────────────── SeuratData v0.2.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mifnb\u001b[39m 3.1.0                                                    \n",
      "\n",
      "\n",
      "────────────────────────────────────────────────────────────── Key ──────────────────────────────────────────────────────────────\n",
      "\n",
      "\u001b[32m✔\u001b[39m Dataset loaded successfully\n",
      "\u001b[33m❯\u001b[39m Dataset built with a newer version of Seurat than installed\n",
      "\u001b[31m❓\u001b[39m Unknown version of Seurat installed\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# row is the gene and columns is the cell in R seurat object.\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratData)\n",
    "library(patchwork)\n",
    "library(SeuratDisk)\n",
    "library(Matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62cc84bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install.packages(\"magrittr\") # package installations are only needed the first time you use it\n",
    "# install.packages(\"dplyr\")    # alternative installation of the %>%\n",
    "library(magrittr) # needs to be run every time you start R and want to use %>%\n",
    "library(dplyr)    # alternatively, this also loads %>%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3a55bdec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: scuttle\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: R6\n",
      "\n",
      "Loading required package: hdf5r\n",
      "\n",
      "\n",
      "Attaching package: ‘hdf5r’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘loomR’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratDisk’:\n",
      "\n",
      "    loom\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install scater https://bioconductor.org/packages/release/bioc/html/scater.html\n",
    "library(scater)\n",
    "# install loomR from GitHub using the remotes package remotes::install_github(repo =\n",
    "# 'mojaveazure/loomR', ref = 'develop')\n",
    "library(loomR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc0b5daa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "               _                           \n",
       "platform       x86_64-pc-linux-gnu         \n",
       "arch           x86_64                      \n",
       "os             linux-gnu                   \n",
       "system         x86_64, linux-gnu           \n",
       "status                                     \n",
       "major          4                           \n",
       "minor          1.1                         \n",
       "year           2021                        \n",
       "month          08                          \n",
       "day            10                          \n",
       "svn rev        80725                       \n",
       "language       R                           \n",
       "version.string R version 4.1.1 (2021-08-10)\n",
       "nickname       Kick Things                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e4683459",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the raw counts data\n",
    "raw_counts <- read.csv(\"count_matrix_pbmc_full.csv\", header= TRUE)\n",
    "\n",
    "# read the meta data\n",
    "cell_meta <- read.csv(\"metadata_cell_pbmc_full.csv\", header= TRUE)\n",
    "gene_meta <- read.csv(\"metadata_gene_pbmc_full.csv\", header= TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ebe76e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create seurat object\n",
    "# mydata <- CreateSeuratObject(counts = raw_counts, min.cells = 3, min.genes = 200, project = \"mydata_scRNAseq\")\n",
    "mydata <- CreateSeuratObject(counts = raw_counts, project = \"mydata_scRNAseq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ce1182",
   "metadata": {},
   "outputs": [],
   "source": [
    "mydata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "448983d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding cell meta data such as batch and cell type\n",
    "mydata$batch <- cell_meta$batch\n",
    "mydata$Method <- cell_meta$Method\n",
    "mydata$CellType <- cell_meta$CellType"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e8d563ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 15476 × 6</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>batch</th><th scope=col>Method</th><th scope=col>CellType</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>X0</th><td>mydata_scRNAseq</td><td> 639.8319</td><td>318</td><td>0</td><td>10X_3prime</td><td>B cell         </td></tr>\n",
       "\t<tr><th scope=row>X1</th><td>mydata_scRNAseq</td><td> 653.7351</td><td>419</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X2</th><td>mydata_scRNAseq</td><td> 482.5800</td><td>255</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X3</th><td>mydata_scRNAseq</td><td>1007.1893</td><td>636</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X4</th><td>mydata_scRNAseq</td><td>1008.2598</td><td>619</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X5</th><td>mydata_scRNAseq</td><td> 726.0322</td><td>401</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X6</th><td>mydata_scRNAseq</td><td> 616.3565</td><td>419</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X7</th><td>mydata_scRNAseq</td><td> 989.5281</td><td>647</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X8</th><td>mydata_scRNAseq</td><td> 553.2230</td><td>352</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X9</th><td>mydata_scRNAseq</td><td> 705.3900</td><td>411</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X10</th><td>mydata_scRNAseq</td><td>1041.8196</td><td>726</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X11</th><td>mydata_scRNAseq</td><td> 664.0299</td><td>471</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X12</th><td>mydata_scRNAseq</td><td> 660.9768</td><td>444</td><td>0</td><td>10X_3prime</td><td>B cell         </td></tr>\n",
       "\t<tr><th scope=row>X13</th><td>mydata_scRNAseq</td><td> 756.6568</td><td>450</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X14</th><td>mydata_scRNAseq</td><td> 580.0542</td><td>409</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X15</th><td>mydata_scRNAseq</td><td> 769.2080</td><td>498</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X16</th><td>mydata_scRNAseq</td><td> 586.9649</td><td>347</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X17</th><td>mydata_scRNAseq</td><td> 731.5836</td><td>520</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X18</th><td>mydata_scRNAseq</td><td> 981.4261</td><td>687</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X19</th><td>mydata_scRNAseq</td><td> 612.8469</td><td>349</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X20</th><td>mydata_scRNAseq</td><td>1048.1081</td><td>869</td><td>0</td><td>10X_3prime</td><td>Monocyte_CD14  </td></tr>\n",
       "\t<tr><th scope=row>X21</th><td>mydata_scRNAseq</td><td> 514.9374</td><td>290</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X22</th><td>mydata_scRNAseq</td><td> 925.0391</td><td>515</td><td>0</td><td>10X_3prime</td><td>Monocyte_FCGR3A</td></tr>\n",
       "\t<tr><th scope=row>X23</th><td>mydata_scRNAseq</td><td> 646.3353</td><td>334</td><td>0</td><td>10X_3prime</td><td>B cell         </td></tr>\n",
       "\t<tr><th scope=row>X24</th><td>mydata_scRNAseq</td><td> 644.8267</td><td>379</td><td>0</td><td>10X_3prime</td><td>CD8 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X25</th><td>mydata_scRNAseq</td><td> 633.4876</td><td>332</td><td>0</td><td>10X_3prime</td><td>B cell         </td></tr>\n",
       "\t<tr><th scope=row>X26</th><td>mydata_scRNAseq</td><td> 873.1821</td><td>507</td><td>0</td><td>10X_3prime</td><td>NK cell        </td></tr>\n",
       "\t<tr><th scope=row>X27</th><td>mydata_scRNAseq</td><td> 769.3820</td><td>548</td><td>0</td><td>10X_3prime</td><td>B cell         </td></tr>\n",
       "\t<tr><th scope=row>X28</th><td>mydata_scRNAseq</td><td> 461.0303</td><td>257</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>X29</th><td>mydata_scRNAseq</td><td> 565.6358</td><td>340</td><td>0</td><td>10X_3prime</td><td>CD4 T cell     </td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>X15446</th><td>mydata_scRNAseq</td><td>1268.4752</td><td>1020</td><td>1</td><td>10X_5prime</td><td>Monocyte_FCGR3A            </td></tr>\n",
       "\t<tr><th scope=row>X15447</th><td>mydata_scRNAseq</td><td> 868.1823</td><td> 766</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15448</th><td>mydata_scRNAseq</td><td> 748.6415</td><td> 547</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15449</th><td>mydata_scRNAseq</td><td> 992.4343</td><td> 683</td><td>1</td><td>10X_5prime</td><td>NK cell                    </td></tr>\n",
       "\t<tr><th scope=row>X15450</th><td>mydata_scRNAseq</td><td> 787.9778</td><td> 424</td><td>1</td><td>10X_5prime</td><td>B cell                     </td></tr>\n",
       "\t<tr><th scope=row>X15451</th><td>mydata_scRNAseq</td><td> 727.4682</td><td> 361</td><td>1</td><td>10X_5prime</td><td>B cell                     </td></tr>\n",
       "\t<tr><th scope=row>X15452</th><td>mydata_scRNAseq</td><td>1260.9854</td><td>1270</td><td>1</td><td>10X_5prime</td><td>Monocyte_CD14              </td></tr>\n",
       "\t<tr><th scope=row>X15453</th><td>mydata_scRNAseq</td><td>1084.0752</td><td> 585</td><td>1</td><td>10X_5prime</td><td>Megakaryocyte              </td></tr>\n",
       "\t<tr><th scope=row>X15454</th><td>mydata_scRNAseq</td><td> 880.4903</td><td> 652</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15455</th><td>mydata_scRNAseq</td><td> 859.7235</td><td> 503</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15456</th><td>mydata_scRNAseq</td><td> 684.9354</td><td> 496</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15457</th><td>mydata_scRNAseq</td><td> 996.4487</td><td> 635</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15458</th><td>mydata_scRNAseq</td><td> 856.6570</td><td> 545</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15459</th><td>mydata_scRNAseq</td><td> 699.8400</td><td> 415</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15460</th><td>mydata_scRNAseq</td><td> 836.5762</td><td> 594</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15461</th><td>mydata_scRNAseq</td><td> 897.0903</td><td> 577</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15462</th><td>mydata_scRNAseq</td><td> 683.7456</td><td> 422</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15463</th><td>mydata_scRNAseq</td><td> 914.6368</td><td> 612</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15464</th><td>mydata_scRNAseq</td><td> 596.7657</td><td> 311</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15465</th><td>mydata_scRNAseq</td><td>1227.4147</td><td> 867</td><td>1</td><td>10X_5prime</td><td>Monocyte_CD14              </td></tr>\n",
       "\t<tr><th scope=row>X15466</th><td>mydata_scRNAseq</td><td> 835.8752</td><td> 550</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15467</th><td>mydata_scRNAseq</td><td> 644.1281</td><td> 442</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15468</th><td>mydata_scRNAseq</td><td> 714.7969</td><td> 516</td><td>1</td><td>10X_5prime</td><td>CD4 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15469</th><td>mydata_scRNAseq</td><td>1190.0775</td><td> 850</td><td>1</td><td>10X_5prime</td><td>Monocyte_CD14              </td></tr>\n",
       "\t<tr><th scope=row>X15470</th><td>mydata_scRNAseq</td><td>1082.6097</td><td>1037</td><td>1</td><td>10X_5prime</td><td>B cell                     </td></tr>\n",
       "\t<tr><th scope=row>X15471</th><td>mydata_scRNAseq</td><td>1162.3105</td><td> 812</td><td>1</td><td>10X_5prime</td><td>Monocyte_CD14              </td></tr>\n",
       "\t<tr><th scope=row>X15472</th><td>mydata_scRNAseq</td><td> 791.7360</td><td> 454</td><td>1</td><td>10X_5prime</td><td>B cell                     </td></tr>\n",
       "\t<tr><th scope=row>X15473</th><td>mydata_scRNAseq</td><td> 806.5532</td><td> 480</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "\t<tr><th scope=row>X15474</th><td>mydata_scRNAseq</td><td>1185.6274</td><td> 877</td><td>1</td><td>10X_5prime</td><td>Plasmacytoid dendritic cell</td></tr>\n",
       "\t<tr><th scope=row>X15475</th><td>mydata_scRNAseq</td><td> 796.2334</td><td> 520</td><td>1</td><td>10X_5prime</td><td>CD8 T cell                 </td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 15476 × 6\n",
       "\\begin{tabular}{r|llllll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & batch & Method & CellType\\\\\n",
       "  & <fct> & <dbl> & <int> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tX0 & mydata\\_scRNAseq &  639.8319 & 318 & 0 & 10X\\_3prime & B cell         \\\\\n",
       "\tX1 & mydata\\_scRNAseq &  653.7351 & 419 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\tX2 & mydata\\_scRNAseq &  482.5800 & 255 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\tX3 & mydata\\_scRNAseq & 1007.1893 & 636 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX4 & mydata\\_scRNAseq & 1008.2598 & 619 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX5 & mydata\\_scRNAseq &  726.0322 & 401 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX6 & mydata\\_scRNAseq &  616.3565 & 419 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX7 & mydata\\_scRNAseq &  989.5281 & 647 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX8 & mydata\\_scRNAseq &  553.2230 & 352 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX9 & mydata\\_scRNAseq &  705.3900 & 411 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\tX10 & mydata\\_scRNAseq & 1041.8196 & 726 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX11 & mydata\\_scRNAseq &  664.0299 & 471 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX12 & mydata\\_scRNAseq &  660.9768 & 444 & 0 & 10X\\_3prime & B cell         \\\\\n",
       "\tX13 & mydata\\_scRNAseq &  756.6568 & 450 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX14 & mydata\\_scRNAseq &  580.0542 & 409 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\tX15 & mydata\\_scRNAseq &  769.2080 & 498 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX16 & mydata\\_scRNAseq &  586.9649 & 347 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX17 & mydata\\_scRNAseq &  731.5836 & 520 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\tX18 & mydata\\_scRNAseq &  981.4261 & 687 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX19 & mydata\\_scRNAseq &  612.8469 & 349 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX20 & mydata\\_scRNAseq & 1048.1081 & 869 & 0 & 10X\\_3prime & Monocyte\\_CD14  \\\\\n",
       "\tX21 & mydata\\_scRNAseq &  514.9374 & 290 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX22 & mydata\\_scRNAseq &  925.0391 & 515 & 0 & 10X\\_3prime & Monocyte\\_FCGR3A\\\\\n",
       "\tX23 & mydata\\_scRNAseq &  646.3353 & 334 & 0 & 10X\\_3prime & B cell         \\\\\n",
       "\tX24 & mydata\\_scRNAseq &  644.8267 & 379 & 0 & 10X\\_3prime & CD8 T cell     \\\\\n",
       "\tX25 & mydata\\_scRNAseq &  633.4876 & 332 & 0 & 10X\\_3prime & B cell         \\\\\n",
       "\tX26 & mydata\\_scRNAseq &  873.1821 & 507 & 0 & 10X\\_3prime & NK cell        \\\\\n",
       "\tX27 & mydata\\_scRNAseq &  769.3820 & 548 & 0 & 10X\\_3prime & B cell         \\\\\n",
       "\tX28 & mydata\\_scRNAseq &  461.0303 & 257 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\tX29 & mydata\\_scRNAseq &  565.6358 & 340 & 0 & 10X\\_3prime & CD4 T cell     \\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tX15446 & mydata\\_scRNAseq & 1268.4752 & 1020 & 1 & 10X\\_5prime & Monocyte\\_FCGR3A            \\\\\n",
       "\tX15447 & mydata\\_scRNAseq &  868.1823 &  766 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15448 & mydata\\_scRNAseq &  748.6415 &  547 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15449 & mydata\\_scRNAseq &  992.4343 &  683 & 1 & 10X\\_5prime & NK cell                    \\\\\n",
       "\tX15450 & mydata\\_scRNAseq &  787.9778 &  424 & 1 & 10X\\_5prime & B cell                     \\\\\n",
       "\tX15451 & mydata\\_scRNAseq &  727.4682 &  361 & 1 & 10X\\_5prime & B cell                     \\\\\n",
       "\tX15452 & mydata\\_scRNAseq & 1260.9854 & 1270 & 1 & 10X\\_5prime & Monocyte\\_CD14              \\\\\n",
       "\tX15453 & mydata\\_scRNAseq & 1084.0752 &  585 & 1 & 10X\\_5prime & Megakaryocyte              \\\\\n",
       "\tX15454 & mydata\\_scRNAseq &  880.4903 &  652 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15455 & mydata\\_scRNAseq &  859.7235 &  503 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15456 & mydata\\_scRNAseq &  684.9354 &  496 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15457 & mydata\\_scRNAseq &  996.4487 &  635 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15458 & mydata\\_scRNAseq &  856.6570 &  545 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15459 & mydata\\_scRNAseq &  699.8400 &  415 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15460 & mydata\\_scRNAseq &  836.5762 &  594 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15461 & mydata\\_scRNAseq &  897.0903 &  577 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15462 & mydata\\_scRNAseq &  683.7456 &  422 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15463 & mydata\\_scRNAseq &  914.6368 &  612 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15464 & mydata\\_scRNAseq &  596.7657 &  311 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15465 & mydata\\_scRNAseq & 1227.4147 &  867 & 1 & 10X\\_5prime & Monocyte\\_CD14              \\\\\n",
       "\tX15466 & mydata\\_scRNAseq &  835.8752 &  550 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15467 & mydata\\_scRNAseq &  644.1281 &  442 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15468 & mydata\\_scRNAseq &  714.7969 &  516 & 1 & 10X\\_5prime & CD4 T cell                 \\\\\n",
       "\tX15469 & mydata\\_scRNAseq & 1190.0775 &  850 & 1 & 10X\\_5prime & Monocyte\\_CD14              \\\\\n",
       "\tX15470 & mydata\\_scRNAseq & 1082.6097 & 1037 & 1 & 10X\\_5prime & B cell                     \\\\\n",
       "\tX15471 & mydata\\_scRNAseq & 1162.3105 &  812 & 1 & 10X\\_5prime & Monocyte\\_CD14              \\\\\n",
       "\tX15472 & mydata\\_scRNAseq &  791.7360 &  454 & 1 & 10X\\_5prime & B cell                     \\\\\n",
       "\tX15473 & mydata\\_scRNAseq &  806.5532 &  480 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\tX15474 & mydata\\_scRNAseq & 1185.6274 &  877 & 1 & 10X\\_5prime & Plasmacytoid dendritic cell\\\\\n",
       "\tX15475 & mydata\\_scRNAseq &  796.2334 &  520 & 1 & 10X\\_5prime & CD8 T cell                 \\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 15476 × 6\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | batch &lt;int&gt; | Method &lt;chr&gt; | CellType &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|---|\n",
       "| X0 | mydata_scRNAseq |  639.8319 | 318 | 0 | 10X_3prime | B cell          |\n",
       "| X1 | mydata_scRNAseq |  653.7351 | 419 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| X2 | mydata_scRNAseq |  482.5800 | 255 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| X3 | mydata_scRNAseq | 1007.1893 | 636 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X4 | mydata_scRNAseq | 1008.2598 | 619 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X5 | mydata_scRNAseq |  726.0322 | 401 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X6 | mydata_scRNAseq |  616.3565 | 419 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X7 | mydata_scRNAseq |  989.5281 | 647 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X8 | mydata_scRNAseq |  553.2230 | 352 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X9 | mydata_scRNAseq |  705.3900 | 411 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| X10 | mydata_scRNAseq | 1041.8196 | 726 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X11 | mydata_scRNAseq |  664.0299 | 471 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X12 | mydata_scRNAseq |  660.9768 | 444 | 0 | 10X_3prime | B cell          |\n",
       "| X13 | mydata_scRNAseq |  756.6568 | 450 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X14 | mydata_scRNAseq |  580.0542 | 409 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| X15 | mydata_scRNAseq |  769.2080 | 498 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X16 | mydata_scRNAseq |  586.9649 | 347 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X17 | mydata_scRNAseq |  731.5836 | 520 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| X18 | mydata_scRNAseq |  981.4261 | 687 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X19 | mydata_scRNAseq |  612.8469 | 349 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X20 | mydata_scRNAseq | 1048.1081 | 869 | 0 | 10X_3prime | Monocyte_CD14   |\n",
       "| X21 | mydata_scRNAseq |  514.9374 | 290 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X22 | mydata_scRNAseq |  925.0391 | 515 | 0 | 10X_3prime | Monocyte_FCGR3A |\n",
       "| X23 | mydata_scRNAseq |  646.3353 | 334 | 0 | 10X_3prime | B cell          |\n",
       "| X24 | mydata_scRNAseq |  644.8267 | 379 | 0 | 10X_3prime | CD8 T cell      |\n",
       "| X25 | mydata_scRNAseq |  633.4876 | 332 | 0 | 10X_3prime | B cell          |\n",
       "| X26 | mydata_scRNAseq |  873.1821 | 507 | 0 | 10X_3prime | NK cell         |\n",
       "| X27 | mydata_scRNAseq |  769.3820 | 548 | 0 | 10X_3prime | B cell          |\n",
       "| X28 | mydata_scRNAseq |  461.0303 | 257 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| X29 | mydata_scRNAseq |  565.6358 | 340 | 0 | 10X_3prime | CD4 T cell      |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| X15446 | mydata_scRNAseq | 1268.4752 | 1020 | 1 | 10X_5prime | Monocyte_FCGR3A             |\n",
       "| X15447 | mydata_scRNAseq |  868.1823 |  766 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15448 | mydata_scRNAseq |  748.6415 |  547 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15449 | mydata_scRNAseq |  992.4343 |  683 | 1 | 10X_5prime | NK cell                     |\n",
       "| X15450 | mydata_scRNAseq |  787.9778 |  424 | 1 | 10X_5prime | B cell                      |\n",
       "| X15451 | mydata_scRNAseq |  727.4682 |  361 | 1 | 10X_5prime | B cell                      |\n",
       "| X15452 | mydata_scRNAseq | 1260.9854 | 1270 | 1 | 10X_5prime | Monocyte_CD14               |\n",
       "| X15453 | mydata_scRNAseq | 1084.0752 |  585 | 1 | 10X_5prime | Megakaryocyte               |\n",
       "| X15454 | mydata_scRNAseq |  880.4903 |  652 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15455 | mydata_scRNAseq |  859.7235 |  503 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15456 | mydata_scRNAseq |  684.9354 |  496 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15457 | mydata_scRNAseq |  996.4487 |  635 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15458 | mydata_scRNAseq |  856.6570 |  545 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15459 | mydata_scRNAseq |  699.8400 |  415 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15460 | mydata_scRNAseq |  836.5762 |  594 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15461 | mydata_scRNAseq |  897.0903 |  577 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15462 | mydata_scRNAseq |  683.7456 |  422 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15463 | mydata_scRNAseq |  914.6368 |  612 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15464 | mydata_scRNAseq |  596.7657 |  311 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15465 | mydata_scRNAseq | 1227.4147 |  867 | 1 | 10X_5prime | Monocyte_CD14               |\n",
       "| X15466 | mydata_scRNAseq |  835.8752 |  550 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15467 | mydata_scRNAseq |  644.1281 |  442 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15468 | mydata_scRNAseq |  714.7969 |  516 | 1 | 10X_5prime | CD4 T cell                  |\n",
       "| X15469 | mydata_scRNAseq | 1190.0775 |  850 | 1 | 10X_5prime | Monocyte_CD14               |\n",
       "| X15470 | mydata_scRNAseq | 1082.6097 | 1037 | 1 | 10X_5prime | B cell                      |\n",
       "| X15471 | mydata_scRNAseq | 1162.3105 |  812 | 1 | 10X_5prime | Monocyte_CD14               |\n",
       "| X15472 | mydata_scRNAseq |  791.7360 |  454 | 1 | 10X_5prime | B cell                      |\n",
       "| X15473 | mydata_scRNAseq |  806.5532 |  480 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "| X15474 | mydata_scRNAseq | 1185.6274 |  877 | 1 | 10X_5prime | Plasmacytoid dendritic cell |\n",
       "| X15475 | mydata_scRNAseq |  796.2334 |  520 | 1 | 10X_5prime | CD8 T cell                  |\n",
       "\n"
      ],
      "text/plain": [
       "       orig.ident      nCount_RNA nFeature_RNA batch Method    \n",
       "X0     mydata_scRNAseq  639.8319  318          0     10X_3prime\n",
       "X1     mydata_scRNAseq  653.7351  419          0     10X_3prime\n",
       "X2     mydata_scRNAseq  482.5800  255          0     10X_3prime\n",
       "X3     mydata_scRNAseq 1007.1893  636          0     10X_3prime\n",
       "X4     mydata_scRNAseq 1008.2598  619          0     10X_3prime\n",
       "X5     mydata_scRNAseq  726.0322  401          0     10X_3prime\n",
       "X6     mydata_scRNAseq  616.3565  419          0     10X_3prime\n",
       "X7     mydata_scRNAseq  989.5281  647          0     10X_3prime\n",
       "X8     mydata_scRNAseq  553.2230  352          0     10X_3prime\n",
       "X9     mydata_scRNAseq  705.3900  411          0     10X_3prime\n",
       "X10    mydata_scRNAseq 1041.8196  726          0     10X_3prime\n",
       "X11    mydata_scRNAseq  664.0299  471          0     10X_3prime\n",
       "X12    mydata_scRNAseq  660.9768  444          0     10X_3prime\n",
       "X13    mydata_scRNAseq  756.6568  450          0     10X_3prime\n",
       "X14    mydata_scRNAseq  580.0542  409          0     10X_3prime\n",
       "X15    mydata_scRNAseq  769.2080  498          0     10X_3prime\n",
       "X16    mydata_scRNAseq  586.9649  347          0     10X_3prime\n",
       "X17    mydata_scRNAseq  731.5836  520          0     10X_3prime\n",
       "X18    mydata_scRNAseq  981.4261  687          0     10X_3prime\n",
       "X19    mydata_scRNAseq  612.8469  349          0     10X_3prime\n",
       "X20    mydata_scRNAseq 1048.1081  869          0     10X_3prime\n",
       "X21    mydata_scRNAseq  514.9374  290          0     10X_3prime\n",
       "X22    mydata_scRNAseq  925.0391  515          0     10X_3prime\n",
       "X23    mydata_scRNAseq  646.3353  334          0     10X_3prime\n",
       "X24    mydata_scRNAseq  644.8267  379          0     10X_3prime\n",
       "X25    mydata_scRNAseq  633.4876  332          0     10X_3prime\n",
       "X26    mydata_scRNAseq  873.1821  507          0     10X_3prime\n",
       "X27    mydata_scRNAseq  769.3820  548          0     10X_3prime\n",
       "X28    mydata_scRNAseq  461.0303  257          0     10X_3prime\n",
       "X29    mydata_scRNAseq  565.6358  340          0     10X_3prime\n",
       "⋮      ⋮               ⋮          ⋮            ⋮     ⋮         \n",
       "X15446 mydata_scRNAseq 1268.4752  1020         1     10X_5prime\n",
       "X15447 mydata_scRNAseq  868.1823   766         1     10X_5prime\n",
       "X15448 mydata_scRNAseq  748.6415   547         1     10X_5prime\n",
       "X15449 mydata_scRNAseq  992.4343   683         1     10X_5prime\n",
       "X15450 mydata_scRNAseq  787.9778   424         1     10X_5prime\n",
       "X15451 mydata_scRNAseq  727.4682   361         1     10X_5prime\n",
       "X15452 mydata_scRNAseq 1260.9854  1270         1     10X_5prime\n",
       "X15453 mydata_scRNAseq 1084.0752   585         1     10X_5prime\n",
       "X15454 mydata_scRNAseq  880.4903   652         1     10X_5prime\n",
       "X15455 mydata_scRNAseq  859.7235   503         1     10X_5prime\n",
       "X15456 mydata_scRNAseq  684.9354   496         1     10X_5prime\n",
       "X15457 mydata_scRNAseq  996.4487   635         1     10X_5prime\n",
       "X15458 mydata_scRNAseq  856.6570   545         1     10X_5prime\n",
       "X15459 mydata_scRNAseq  699.8400   415         1     10X_5prime\n",
       "X15460 mydata_scRNAseq  836.5762   594         1     10X_5prime\n",
       "X15461 mydata_scRNAseq  897.0903   577         1     10X_5prime\n",
       "X15462 mydata_scRNAseq  683.7456   422         1     10X_5prime\n",
       "X15463 mydata_scRNAseq  914.6368   612         1     10X_5prime\n",
       "X15464 mydata_scRNAseq  596.7657   311         1     10X_5prime\n",
       "X15465 mydata_scRNAseq 1227.4147   867         1     10X_5prime\n",
       "X15466 mydata_scRNAseq  835.8752   550         1     10X_5prime\n",
       "X15467 mydata_scRNAseq  644.1281   442         1     10X_5prime\n",
       "X15468 mydata_scRNAseq  714.7969   516         1     10X_5prime\n",
       "X15469 mydata_scRNAseq 1190.0775   850         1     10X_5prime\n",
       "X15470 mydata_scRNAseq 1082.6097  1037         1     10X_5prime\n",
       "X15471 mydata_scRNAseq 1162.3105   812         1     10X_5prime\n",
       "X15472 mydata_scRNAseq  791.7360   454         1     10X_5prime\n",
       "X15473 mydata_scRNAseq  806.5532   480         1     10X_5prime\n",
       "X15474 mydata_scRNAseq 1185.6274   877         1     10X_5prime\n",
       "X15475 mydata_scRNAseq  796.2334   520         1     10X_5prime\n",
       "       CellType                   \n",
       "X0     B cell                     \n",
       "X1     CD4 T cell                 \n",
       "X2     CD4 T cell                 \n",
       "X3     Monocyte_CD14              \n",
       "X4     Monocyte_CD14              \n",
       "X5     CD8 T cell                 \n",
       "X6     CD8 T cell                 \n",
       "X7     Monocyte_CD14              \n",
       "X8     CD8 T cell                 \n",
       "X9     CD4 T cell                 \n",
       "X10    Monocyte_CD14              \n",
       "X11    CD8 T cell                 \n",
       "X12    B cell                     \n",
       "X13    CD8 T cell                 \n",
       "X14    CD4 T cell                 \n",
       "X15    Monocyte_CD14              \n",
       "X16    CD8 T cell                 \n",
       "X17    CD4 T cell                 \n",
       "X18    Monocyte_CD14              \n",
       "X19    CD8 T cell                 \n",
       "X20    Monocyte_CD14              \n",
       "X21    CD8 T cell                 \n",
       "X22    Monocyte_FCGR3A            \n",
       "X23    B cell                     \n",
       "X24    CD8 T cell                 \n",
       "X25    B cell                     \n",
       "X26    NK cell                    \n",
       "X27    B cell                     \n",
       "X28    CD4 T cell                 \n",
       "X29    CD4 T cell                 \n",
       "⋮      ⋮                          \n",
       "X15446 Monocyte_FCGR3A            \n",
       "X15447 CD4 T cell                 \n",
       "X15448 CD8 T cell                 \n",
       "X15449 NK cell                    \n",
       "X15450 B cell                     \n",
       "X15451 B cell                     \n",
       "X15452 Monocyte_CD14              \n",
       "X15453 Megakaryocyte              \n",
       "X15454 CD4 T cell                 \n",
       "X15455 CD8 T cell                 \n",
       "X15456 CD4 T cell                 \n",
       "X15457 CD8 T cell                 \n",
       "X15458 CD8 T cell                 \n",
       "X15459 CD4 T cell                 \n",
       "X15460 CD4 T cell                 \n",
       "X15461 CD8 T cell                 \n",
       "X15462 CD8 T cell                 \n",
       "X15463 CD8 T cell                 \n",
       "X15464 CD4 T cell                 \n",
       "X15465 Monocyte_CD14              \n",
       "X15466 CD8 T cell                 \n",
       "X15467 CD8 T cell                 \n",
       "X15468 CD4 T cell                 \n",
       "X15469 Monocyte_CD14              \n",
       "X15470 B cell                     \n",
       "X15471 Monocyte_CD14              \n",
       "X15472 B cell                     \n",
       "X15473 CD8 T cell                 \n",
       "X15474 Plasmacytoid dendritic cell\n",
       "X15475 CD8 T cell                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mydata[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "600a520b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(mydata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d8a29c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split the dataset into a list of batch object\n",
    "data.list <- SplitObject(mydata, split.by = \"batch\")\n",
    "\n",
    "# normalize and identify variable features for each dataset independently\n",
    "data.list <- lapply(X = data.list, FUN = function(x) {\n",
    "#     x <- NormalizeData(x)\n",
    "    x <- FindVariableFeatures(x, nfeatures = 10000)\n",
    "})\n",
    "\n",
    "# select features that are repeatedly variable across datasets for integration\n",
    "features <- SelectIntegrationFeatures(object.list = data.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b0accb34",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scaling features for provided objects\n",
      "\n",
      "Finding all pairwise anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 16895 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 8673 anchors\n",
      "\n"
     ]
    }
   ],
   "source": [
    "immune.anchors <- FindIntegrationAnchors(object.list = data.list, anchor.features = features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37eff971",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Merging dataset 2 into 1\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_features <- lapply(data.list, row.names) %>% Reduce(intersect, .) \n",
    "immune.combined <- IntegrateData( \n",
    "  anchorset = immune.anchors, \n",
    "  features.to.integrate = all_features,\n",
    "  verbose = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f85bf9f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# immune.combined <- IntegrateData(anchorset = immune.anchors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "59d4a662",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify that we will perform downstream analysis on the corrected data note that the\n",
    "# original unmodified data still resides in the 'RNA' assay\n",
    "# DefaultAssay(immune.combined) <- \"integrated\"\n",
    "\n",
    "# # Run the standard workflow for visualization and clustering\n",
    "# immune.combined <- ScaleData(immune.combined, verbose = FALSE)\n",
    "# immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)\n",
    "# immune.combined <- RunUMAP(immune.combined, reduction = \"pca\", dims = 1:30)\n",
    "# immune.combined <- FindNeighbors(immune.combined, reduction = \"pca\", dims = 1:30)\n",
    "# immune.combined <- FindClusters(immune.combined, resolution = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e98561e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualization\n",
    "# p1 <- DimPlot(immune.combined, reduction = \"umap\", group.by = \"batch\")\n",
    "# p2 <- DimPlot(immune.combined, reduction = \"umap\", label = TRUE, repel = TRUE)\n",
    "# p1 + p2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e8e12c47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assay data with 10000 features for 15476 cells\n",
       "Top 10 variable features:\n",
       " 6313, 583, 581, 2926, 3210, 8772, 9272, 3212, 587, 1242 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "immune.combined$integrated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a947d5d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Overwriting previous file seurat_pbmc.loom”\n",
      "Adding col attribute CellID\n",
      "\n",
      "Adding col attribute orig.ident\n",
      "\n",
      "Adding col attribute nCount_RNA\n",
      "\n",
      "Adding col attribute nFeature_RNA\n",
      "\n",
      "Adding col attribute batch\n",
      "\n",
      "Adding col attribute Method\n",
      "\n",
      "Adding col attribute CellType\n",
      "\n",
      "Adding row attribute Gene\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Class: loom\n",
       "Filename: /disco_500t/rom/DisentangledContrastive/contrastive-sc/notebook/seurat_pbmc.loom\n",
       "Access type: H5F_ACC_RDWR\n",
       "Listing:\n",
       "       name    obj_type  dataset.dims dataset.type_class\n",
       "      attrs   H5I_GROUP          <NA>               <NA>\n",
       "  col_attrs   H5I_GROUP          <NA>               <NA>\n",
       " col_graphs   H5I_GROUP          <NA>               <NA>\n",
       "     layers   H5I_GROUP          <NA>               <NA>\n",
       "     matrix H5I_DATASET 15476 x 10000          H5T_FLOAT\n",
       "  row_attrs   H5I_GROUP          <NA>               <NA>\n",
       " row_graphs   H5I_GROUP          <NA>               <NA>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pancreatic.loom <- as.loom(immune.combined, filename = \"seurat_pbmc.loom\", verbose = FALSE, overwrite = TRUE)\n",
    "pancreatic.loom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0e672cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Always remember to close loom files when done\n",
    "pancreatic.loom$close_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b7ab4324",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
