{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "5ce1cd3e",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Attaching SeuratObject\n",
      "\n",
      "── \u001b[1mInstalled datasets\u001b[22m ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── SeuratData v0.2.1 ──\n",
      "\n",
      "\u001b[32m✔\u001b[39m \u001b[34mifnb\u001b[39m 3.1.0                                                                                                                  \n",
      "\n",
      "\n",
      "──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── Key ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────\n",
      "\n",
      "\u001b[32m✔\u001b[39m Dataset loaded successfully\n",
      "\u001b[33m❯\u001b[39m Dataset built with a newer version of Seurat than installed\n",
      "\u001b[31m❓\u001b[39m Unknown version of Seurat installed\n",
      "\n",
      "\n",
      "Registered S3 method overwritten by 'SeuratDisk':\n",
      "  method            from  \n",
      "  as.sparse.H5Group Seurat\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# row is the gene and columns is the cell in R seurat object.\n",
    "\n",
    "library(Seurat)\n",
    "library(SeuratData)\n",
    "library(patchwork)\n",
    "library(SeuratDisk)\n",
    "library(Matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "62cc84bb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "Attaching package: ‘dplyr’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install.packages(\"magrittr\") # package installations are only needed the first time you use it\n",
    "# install.packages(\"dplyr\")    # alternative installation of the %>%\n",
    "library(magrittr) # needs to be run every time you start R and want to use %>%\n",
    "library(dplyr)    # alternatively, this also loads %>%"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "3a55bdec",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: SingleCellExperiment\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: ‘matrixStats’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: ‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, append, as.data.frame, basename, cbind, colnames,\n",
      "    dirname, do.call, duplicated, eval, evalq, Filter, Find, get, grep,\n",
      "    grepl, intersect, is.unsorted, lapply, Map, mapply, match, mget,\n",
      "    order, paste, pmax, pmax.int, pmin, pmin.int, Position, rank,\n",
      "    rbind, Reduce, rownames, sapply, setdiff, sort, table, tapply,\n",
      "    union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: ‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    first, rename\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:Matrix’:\n",
      "\n",
      "    expand, unname\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "\n",
      "Attaching package: ‘IRanges’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:dplyr’:\n",
      "\n",
      "    collapse, desc, slice\n",
      "\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "Loading required package: scuttle\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: R6\n",
      "\n",
      "Loading required package: hdf5r\n",
      "\n",
      "\n",
      "Attaching package: ‘hdf5r’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SummarizedExperiment’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:S4Vectors’:\n",
      "\n",
      "    values\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: ‘loomR’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:dplyr’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratDisk’:\n",
      "\n",
      "    loom\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "# install scater https://bioconductor.org/packages/release/bioc/html/scater.html\n",
    "library(scater)\n",
    "# install loomR from GitHub using the remotes package remotes::install_github(repo =\n",
    "# 'mojaveazure/loomR', ref = 'develop')\n",
    "library(loomR)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "bc0b5daa",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "               _                           \n",
       "platform       x86_64-pc-linux-gnu         \n",
       "arch           x86_64                      \n",
       "os             linux-gnu                   \n",
       "system         x86_64, linux-gnu           \n",
       "status                                     \n",
       "major          4                           \n",
       "minor          1.1                         \n",
       "year           2021                        \n",
       "month          08                          \n",
       "day            10                          \n",
       "svn rev        80725                       \n",
       "language       R                           \n",
       "version.string R version 4.1.1 (2021-08-10)\n",
       "nickname       Kick Things                 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "version"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "e4683459",
   "metadata": {},
   "outputs": [],
   "source": [
    "# read the raw counts data\n",
    "raw_counts <- read.csv(\"simul1_dropout_47_b1_1000_b2_2000/counts.txt\", sep=\"\\t\", header= TRUE)\n",
    "\n",
    "# read the meta data\n",
    "cell_meta <- read.csv(\"simul1_dropout_47_b1_1000_b2_2000/cellinfo.txt\", sep=\"\\t\", header= TRUE)\n",
    "gene_meta <- read.csv(\"simul1_dropout_47_b1_1000_b2_2000/geneinfo.txt\", sep=\"\\t\", header= TRUE)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ebe76e87",
   "metadata": {},
   "outputs": [],
   "source": [
    "# create seurat object\n",
    "# mydata <- CreateSeuratObject(counts = raw_counts, min.cells = 3, min.genes = 200, project = \"mydata_scRNAseq\")\n",
    "mydata <- CreateSeuratObject(counts = raw_counts, project = \"mydata_scRNAseq\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b7ce1182",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "An object of class Seurat \n",
       "5000 features across 8000 samples within 1 assay \n",
       "Active assay: RNA (5000 features, 0 variable features)"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mydata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "448983d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# adding cell meta data such as batch and cell type\n",
    "mydata$Batch <- cell_meta$Batch\n",
    "mydata$Group <- cell_meta$Group"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e8d563ae",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 8000 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>orig.ident</th><th scope=col>nCount_RNA</th><th scope=col>nFeature_RNA</th><th scope=col>Batch</th><th scope=col>Group</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;fct&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;int&gt;</th><th scope=col>&lt;chr&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>Cell1</th><td>mydata_scRNAseq</td><td>42111</td><td>378</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell2</th><td>mydata_scRNAseq</td><td>17649</td><td>229</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell3</th><td>mydata_scRNAseq</td><td>32522</td><td>345</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell4</th><td>mydata_scRNAseq</td><td>22980</td><td>261</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell5</th><td>mydata_scRNAseq</td><td>15843</td><td>217</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell6</th><td>mydata_scRNAseq</td><td>29653</td><td>307</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7</th><td>mydata_scRNAseq</td><td>27245</td><td>275</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell8</th><td>mydata_scRNAseq</td><td>20209</td><td>253</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell9</th><td>mydata_scRNAseq</td><td>34618</td><td>335</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell10</th><td>mydata_scRNAseq</td><td>27335</td><td>304</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell11</th><td>mydata_scRNAseq</td><td>27524</td><td>294</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell12</th><td>mydata_scRNAseq</td><td>14837</td><td>191</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell13</th><td>mydata_scRNAseq</td><td>20973</td><td>231</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell14</th><td>mydata_scRNAseq</td><td>22169</td><td>283</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell15</th><td>mydata_scRNAseq</td><td>31974</td><td>325</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell16</th><td>mydata_scRNAseq</td><td>28446</td><td>306</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell17</th><td>mydata_scRNAseq</td><td>21714</td><td>261</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell18</th><td>mydata_scRNAseq</td><td>18844</td><td>248</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell19</th><td>mydata_scRNAseq</td><td>31582</td><td>312</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell20</th><td>mydata_scRNAseq</td><td>16314</td><td>203</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell21</th><td>mydata_scRNAseq</td><td>25230</td><td>268</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell22</th><td>mydata_scRNAseq</td><td>30230</td><td>295</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell23</th><td>mydata_scRNAseq</td><td>35999</td><td>326</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell24</th><td>mydata_scRNAseq</td><td>32908</td><td>316</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell25</th><td>mydata_scRNAseq</td><td>23052</td><td>263</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell26</th><td>mydata_scRNAseq</td><td>43089</td><td>368</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell27</th><td>mydata_scRNAseq</td><td>29201</td><td>300</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell28</th><td>mydata_scRNAseq</td><td>26147</td><td>280</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell29</th><td>mydata_scRNAseq</td><td>20968</td><td>248</td><td>Batch1</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell30</th><td>mydata_scRNAseq</td><td>22366</td><td>263</td><td>Batch1</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>⋮</th><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td><td>⋮</td></tr>\n",
       "\t<tr><th scope=row>Cell7971</th><td>mydata_scRNAseq</td><td>36476</td><td>339</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7972</th><td>mydata_scRNAseq</td><td>35911</td><td>315</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7973</th><td>mydata_scRNAseq</td><td>21574</td><td>255</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7974</th><td>mydata_scRNAseq</td><td>12153</td><td>184</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7975</th><td>mydata_scRNAseq</td><td>23563</td><td>266</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7976</th><td>mydata_scRNAseq</td><td>26029</td><td>266</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7977</th><td>mydata_scRNAseq</td><td>34293</td><td>297</td><td>Batch4</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell7978</th><td>mydata_scRNAseq</td><td>24028</td><td>268</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7979</th><td>mydata_scRNAseq</td><td>19475</td><td>223</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7980</th><td>mydata_scRNAseq</td><td>20818</td><td>234</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7981</th><td>mydata_scRNAseq</td><td>18762</td><td>213</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7982</th><td>mydata_scRNAseq</td><td>36168</td><td>316</td><td>Batch4</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell7983</th><td>mydata_scRNAseq</td><td>31169</td><td>303</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7984</th><td>mydata_scRNAseq</td><td>38767</td><td>324</td><td>Batch4</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell7985</th><td>mydata_scRNAseq</td><td>40214</td><td>339</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7986</th><td>mydata_scRNAseq</td><td>25817</td><td>266</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7987</th><td>mydata_scRNAseq</td><td>22567</td><td>249</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7988</th><td>mydata_scRNAseq</td><td>28380</td><td>283</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7989</th><td>mydata_scRNAseq</td><td>32846</td><td>297</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7990</th><td>mydata_scRNAseq</td><td>30599</td><td>285</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7991</th><td>mydata_scRNAseq</td><td>57314</td><td>391</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7992</th><td>mydata_scRNAseq</td><td>26472</td><td>253</td><td>Batch4</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell7993</th><td>mydata_scRNAseq</td><td>44935</td><td>385</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7994</th><td>mydata_scRNAseq</td><td>39253</td><td>328</td><td>Batch4</td><td>Group2</td></tr>\n",
       "\t<tr><th scope=row>Cell7995</th><td>mydata_scRNAseq</td><td>27576</td><td>287</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7996</th><td>mydata_scRNAseq</td><td>16254</td><td>195</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7997</th><td>mydata_scRNAseq</td><td>42086</td><td>344</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7998</th><td>mydata_scRNAseq</td><td>37344</td><td>331</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell7999</th><td>mydata_scRNAseq</td><td>25708</td><td>269</td><td>Batch4</td><td>Group1</td></tr>\n",
       "\t<tr><th scope=row>Cell8000</th><td>mydata_scRNAseq</td><td>21201</td><td>225</td><td>Batch4</td><td>Group2</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 8000 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & orig.ident & nCount\\_RNA & nFeature\\_RNA & Batch & Group\\\\\n",
       "  & <fct> & <dbl> & <int> & <chr> & <chr>\\\\\n",
       "\\hline\n",
       "\tCell1 & mydata\\_scRNAseq & 42111 & 378 & Batch1 & Group1\\\\\n",
       "\tCell2 & mydata\\_scRNAseq & 17649 & 229 & Batch1 & Group2\\\\\n",
       "\tCell3 & mydata\\_scRNAseq & 32522 & 345 & Batch1 & Group1\\\\\n",
       "\tCell4 & mydata\\_scRNAseq & 22980 & 261 & Batch1 & Group1\\\\\n",
       "\tCell5 & mydata\\_scRNAseq & 15843 & 217 & Batch1 & Group1\\\\\n",
       "\tCell6 & mydata\\_scRNAseq & 29653 & 307 & Batch1 & Group1\\\\\n",
       "\tCell7 & mydata\\_scRNAseq & 27245 & 275 & Batch1 & Group1\\\\\n",
       "\tCell8 & mydata\\_scRNAseq & 20209 & 253 & Batch1 & Group1\\\\\n",
       "\tCell9 & mydata\\_scRNAseq & 34618 & 335 & Batch1 & Group1\\\\\n",
       "\tCell10 & mydata\\_scRNAseq & 27335 & 304 & Batch1 & Group1\\\\\n",
       "\tCell11 & mydata\\_scRNAseq & 27524 & 294 & Batch1 & Group1\\\\\n",
       "\tCell12 & mydata\\_scRNAseq & 14837 & 191 & Batch1 & Group1\\\\\n",
       "\tCell13 & mydata\\_scRNAseq & 20973 & 231 & Batch1 & Group2\\\\\n",
       "\tCell14 & mydata\\_scRNAseq & 22169 & 283 & Batch1 & Group2\\\\\n",
       "\tCell15 & mydata\\_scRNAseq & 31974 & 325 & Batch1 & Group2\\\\\n",
       "\tCell16 & mydata\\_scRNAseq & 28446 & 306 & Batch1 & Group2\\\\\n",
       "\tCell17 & mydata\\_scRNAseq & 21714 & 261 & Batch1 & Group1\\\\\n",
       "\tCell18 & mydata\\_scRNAseq & 18844 & 248 & Batch1 & Group1\\\\\n",
       "\tCell19 & mydata\\_scRNAseq & 31582 & 312 & Batch1 & Group1\\\\\n",
       "\tCell20 & mydata\\_scRNAseq & 16314 & 203 & Batch1 & Group1\\\\\n",
       "\tCell21 & mydata\\_scRNAseq & 25230 & 268 & Batch1 & Group2\\\\\n",
       "\tCell22 & mydata\\_scRNAseq & 30230 & 295 & Batch1 & Group1\\\\\n",
       "\tCell23 & mydata\\_scRNAseq & 35999 & 326 & Batch1 & Group1\\\\\n",
       "\tCell24 & mydata\\_scRNAseq & 32908 & 316 & Batch1 & Group2\\\\\n",
       "\tCell25 & mydata\\_scRNAseq & 23052 & 263 & Batch1 & Group2\\\\\n",
       "\tCell26 & mydata\\_scRNAseq & 43089 & 368 & Batch1 & Group1\\\\\n",
       "\tCell27 & mydata\\_scRNAseq & 29201 & 300 & Batch1 & Group1\\\\\n",
       "\tCell28 & mydata\\_scRNAseq & 26147 & 280 & Batch1 & Group1\\\\\n",
       "\tCell29 & mydata\\_scRNAseq & 20968 & 248 & Batch1 & Group2\\\\\n",
       "\tCell30 & mydata\\_scRNAseq & 22366 & 263 & Batch1 & Group1\\\\\n",
       "\t⋮ & ⋮ & ⋮ & ⋮ & ⋮ & ⋮\\\\\n",
       "\tCell7971 & mydata\\_scRNAseq & 36476 & 339 & Batch4 & Group1\\\\\n",
       "\tCell7972 & mydata\\_scRNAseq & 35911 & 315 & Batch4 & Group1\\\\\n",
       "\tCell7973 & mydata\\_scRNAseq & 21574 & 255 & Batch4 & Group1\\\\\n",
       "\tCell7974 & mydata\\_scRNAseq & 12153 & 184 & Batch4 & Group1\\\\\n",
       "\tCell7975 & mydata\\_scRNAseq & 23563 & 266 & Batch4 & Group1\\\\\n",
       "\tCell7976 & mydata\\_scRNAseq & 26029 & 266 & Batch4 & Group1\\\\\n",
       "\tCell7977 & mydata\\_scRNAseq & 34293 & 297 & Batch4 & Group2\\\\\n",
       "\tCell7978 & mydata\\_scRNAseq & 24028 & 268 & Batch4 & Group1\\\\\n",
       "\tCell7979 & mydata\\_scRNAseq & 19475 & 223 & Batch4 & Group1\\\\\n",
       "\tCell7980 & mydata\\_scRNAseq & 20818 & 234 & Batch4 & Group1\\\\\n",
       "\tCell7981 & mydata\\_scRNAseq & 18762 & 213 & Batch4 & Group1\\\\\n",
       "\tCell7982 & mydata\\_scRNAseq & 36168 & 316 & Batch4 & Group2\\\\\n",
       "\tCell7983 & mydata\\_scRNAseq & 31169 & 303 & Batch4 & Group1\\\\\n",
       "\tCell7984 & mydata\\_scRNAseq & 38767 & 324 & Batch4 & Group2\\\\\n",
       "\tCell7985 & mydata\\_scRNAseq & 40214 & 339 & Batch4 & Group1\\\\\n",
       "\tCell7986 & mydata\\_scRNAseq & 25817 & 266 & Batch4 & Group1\\\\\n",
       "\tCell7987 & mydata\\_scRNAseq & 22567 & 249 & Batch4 & Group1\\\\\n",
       "\tCell7988 & mydata\\_scRNAseq & 28380 & 283 & Batch4 & Group1\\\\\n",
       "\tCell7989 & mydata\\_scRNAseq & 32846 & 297 & Batch4 & Group1\\\\\n",
       "\tCell7990 & mydata\\_scRNAseq & 30599 & 285 & Batch4 & Group1\\\\\n",
       "\tCell7991 & mydata\\_scRNAseq & 57314 & 391 & Batch4 & Group1\\\\\n",
       "\tCell7992 & mydata\\_scRNAseq & 26472 & 253 & Batch4 & Group2\\\\\n",
       "\tCell7993 & mydata\\_scRNAseq & 44935 & 385 & Batch4 & Group1\\\\\n",
       "\tCell7994 & mydata\\_scRNAseq & 39253 & 328 & Batch4 & Group2\\\\\n",
       "\tCell7995 & mydata\\_scRNAseq & 27576 & 287 & Batch4 & Group1\\\\\n",
       "\tCell7996 & mydata\\_scRNAseq & 16254 & 195 & Batch4 & Group1\\\\\n",
       "\tCell7997 & mydata\\_scRNAseq & 42086 & 344 & Batch4 & Group1\\\\\n",
       "\tCell7998 & mydata\\_scRNAseq & 37344 & 331 & Batch4 & Group1\\\\\n",
       "\tCell7999 & mydata\\_scRNAseq & 25708 & 269 & Batch4 & Group1\\\\\n",
       "\tCell8000 & mydata\\_scRNAseq & 21201 & 225 & Batch4 & Group2\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 8000 × 5\n",
       "\n",
       "| <!--/--> | orig.ident &lt;fct&gt; | nCount_RNA &lt;dbl&gt; | nFeature_RNA &lt;int&gt; | Batch &lt;chr&gt; | Group &lt;chr&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| Cell1 | mydata_scRNAseq | 42111 | 378 | Batch1 | Group1 |\n",
       "| Cell2 | mydata_scRNAseq | 17649 | 229 | Batch1 | Group2 |\n",
       "| Cell3 | mydata_scRNAseq | 32522 | 345 | Batch1 | Group1 |\n",
       "| Cell4 | mydata_scRNAseq | 22980 | 261 | Batch1 | Group1 |\n",
       "| Cell5 | mydata_scRNAseq | 15843 | 217 | Batch1 | Group1 |\n",
       "| Cell6 | mydata_scRNAseq | 29653 | 307 | Batch1 | Group1 |\n",
       "| Cell7 | mydata_scRNAseq | 27245 | 275 | Batch1 | Group1 |\n",
       "| Cell8 | mydata_scRNAseq | 20209 | 253 | Batch1 | Group1 |\n",
       "| Cell9 | mydata_scRNAseq | 34618 | 335 | Batch1 | Group1 |\n",
       "| Cell10 | mydata_scRNAseq | 27335 | 304 | Batch1 | Group1 |\n",
       "| Cell11 | mydata_scRNAseq | 27524 | 294 | Batch1 | Group1 |\n",
       "| Cell12 | mydata_scRNAseq | 14837 | 191 | Batch1 | Group1 |\n",
       "| Cell13 | mydata_scRNAseq | 20973 | 231 | Batch1 | Group2 |\n",
       "| Cell14 | mydata_scRNAseq | 22169 | 283 | Batch1 | Group2 |\n",
       "| Cell15 | mydata_scRNAseq | 31974 | 325 | Batch1 | Group2 |\n",
       "| Cell16 | mydata_scRNAseq | 28446 | 306 | Batch1 | Group2 |\n",
       "| Cell17 | mydata_scRNAseq | 21714 | 261 | Batch1 | Group1 |\n",
       "| Cell18 | mydata_scRNAseq | 18844 | 248 | Batch1 | Group1 |\n",
       "| Cell19 | mydata_scRNAseq | 31582 | 312 | Batch1 | Group1 |\n",
       "| Cell20 | mydata_scRNAseq | 16314 | 203 | Batch1 | Group1 |\n",
       "| Cell21 | mydata_scRNAseq | 25230 | 268 | Batch1 | Group2 |\n",
       "| Cell22 | mydata_scRNAseq | 30230 | 295 | Batch1 | Group1 |\n",
       "| Cell23 | mydata_scRNAseq | 35999 | 326 | Batch1 | Group1 |\n",
       "| Cell24 | mydata_scRNAseq | 32908 | 316 | Batch1 | Group2 |\n",
       "| Cell25 | mydata_scRNAseq | 23052 | 263 | Batch1 | Group2 |\n",
       "| Cell26 | mydata_scRNAseq | 43089 | 368 | Batch1 | Group1 |\n",
       "| Cell27 | mydata_scRNAseq | 29201 | 300 | Batch1 | Group1 |\n",
       "| Cell28 | mydata_scRNAseq | 26147 | 280 | Batch1 | Group1 |\n",
       "| Cell29 | mydata_scRNAseq | 20968 | 248 | Batch1 | Group2 |\n",
       "| Cell30 | mydata_scRNAseq | 22366 | 263 | Batch1 | Group1 |\n",
       "| ⋮ | ⋮ | ⋮ | ⋮ | ⋮ | ⋮ |\n",
       "| Cell7971 | mydata_scRNAseq | 36476 | 339 | Batch4 | Group1 |\n",
       "| Cell7972 | mydata_scRNAseq | 35911 | 315 | Batch4 | Group1 |\n",
       "| Cell7973 | mydata_scRNAseq | 21574 | 255 | Batch4 | Group1 |\n",
       "| Cell7974 | mydata_scRNAseq | 12153 | 184 | Batch4 | Group1 |\n",
       "| Cell7975 | mydata_scRNAseq | 23563 | 266 | Batch4 | Group1 |\n",
       "| Cell7976 | mydata_scRNAseq | 26029 | 266 | Batch4 | Group1 |\n",
       "| Cell7977 | mydata_scRNAseq | 34293 | 297 | Batch4 | Group2 |\n",
       "| Cell7978 | mydata_scRNAseq | 24028 | 268 | Batch4 | Group1 |\n",
       "| Cell7979 | mydata_scRNAseq | 19475 | 223 | Batch4 | Group1 |\n",
       "| Cell7980 | mydata_scRNAseq | 20818 | 234 | Batch4 | Group1 |\n",
       "| Cell7981 | mydata_scRNAseq | 18762 | 213 | Batch4 | Group1 |\n",
       "| Cell7982 | mydata_scRNAseq | 36168 | 316 | Batch4 | Group2 |\n",
       "| Cell7983 | mydata_scRNAseq | 31169 | 303 | Batch4 | Group1 |\n",
       "| Cell7984 | mydata_scRNAseq | 38767 | 324 | Batch4 | Group2 |\n",
       "| Cell7985 | mydata_scRNAseq | 40214 | 339 | Batch4 | Group1 |\n",
       "| Cell7986 | mydata_scRNAseq | 25817 | 266 | Batch4 | Group1 |\n",
       "| Cell7987 | mydata_scRNAseq | 22567 | 249 | Batch4 | Group1 |\n",
       "| Cell7988 | mydata_scRNAseq | 28380 | 283 | Batch4 | Group1 |\n",
       "| Cell7989 | mydata_scRNAseq | 32846 | 297 | Batch4 | Group1 |\n",
       "| Cell7990 | mydata_scRNAseq | 30599 | 285 | Batch4 | Group1 |\n",
       "| Cell7991 | mydata_scRNAseq | 57314 | 391 | Batch4 | Group1 |\n",
       "| Cell7992 | mydata_scRNAseq | 26472 | 253 | Batch4 | Group2 |\n",
       "| Cell7993 | mydata_scRNAseq | 44935 | 385 | Batch4 | Group1 |\n",
       "| Cell7994 | mydata_scRNAseq | 39253 | 328 | Batch4 | Group2 |\n",
       "| Cell7995 | mydata_scRNAseq | 27576 | 287 | Batch4 | Group1 |\n",
       "| Cell7996 | mydata_scRNAseq | 16254 | 195 | Batch4 | Group1 |\n",
       "| Cell7997 | mydata_scRNAseq | 42086 | 344 | Batch4 | Group1 |\n",
       "| Cell7998 | mydata_scRNAseq | 37344 | 331 | Batch4 | Group1 |\n",
       "| Cell7999 | mydata_scRNAseq | 25708 | 269 | Batch4 | Group1 |\n",
       "| Cell8000 | mydata_scRNAseq | 21201 | 225 | Batch4 | Group2 |\n",
       "\n"
      ],
      "text/plain": [
       "         orig.ident      nCount_RNA nFeature_RNA Batch  Group \n",
       "Cell1    mydata_scRNAseq 42111      378          Batch1 Group1\n",
       "Cell2    mydata_scRNAseq 17649      229          Batch1 Group2\n",
       "Cell3    mydata_scRNAseq 32522      345          Batch1 Group1\n",
       "Cell4    mydata_scRNAseq 22980      261          Batch1 Group1\n",
       "Cell5    mydata_scRNAseq 15843      217          Batch1 Group1\n",
       "Cell6    mydata_scRNAseq 29653      307          Batch1 Group1\n",
       "Cell7    mydata_scRNAseq 27245      275          Batch1 Group1\n",
       "Cell8    mydata_scRNAseq 20209      253          Batch1 Group1\n",
       "Cell9    mydata_scRNAseq 34618      335          Batch1 Group1\n",
       "Cell10   mydata_scRNAseq 27335      304          Batch1 Group1\n",
       "Cell11   mydata_scRNAseq 27524      294          Batch1 Group1\n",
       "Cell12   mydata_scRNAseq 14837      191          Batch1 Group1\n",
       "Cell13   mydata_scRNAseq 20973      231          Batch1 Group2\n",
       "Cell14   mydata_scRNAseq 22169      283          Batch1 Group2\n",
       "Cell15   mydata_scRNAseq 31974      325          Batch1 Group2\n",
       "Cell16   mydata_scRNAseq 28446      306          Batch1 Group2\n",
       "Cell17   mydata_scRNAseq 21714      261          Batch1 Group1\n",
       "Cell18   mydata_scRNAseq 18844      248          Batch1 Group1\n",
       "Cell19   mydata_scRNAseq 31582      312          Batch1 Group1\n",
       "Cell20   mydata_scRNAseq 16314      203          Batch1 Group1\n",
       "Cell21   mydata_scRNAseq 25230      268          Batch1 Group2\n",
       "Cell22   mydata_scRNAseq 30230      295          Batch1 Group1\n",
       "Cell23   mydata_scRNAseq 35999      326          Batch1 Group1\n",
       "Cell24   mydata_scRNAseq 32908      316          Batch1 Group2\n",
       "Cell25   mydata_scRNAseq 23052      263          Batch1 Group2\n",
       "Cell26   mydata_scRNAseq 43089      368          Batch1 Group1\n",
       "Cell27   mydata_scRNAseq 29201      300          Batch1 Group1\n",
       "Cell28   mydata_scRNAseq 26147      280          Batch1 Group1\n",
       "Cell29   mydata_scRNAseq 20968      248          Batch1 Group2\n",
       "Cell30   mydata_scRNAseq 22366      263          Batch1 Group1\n",
       "⋮        ⋮               ⋮          ⋮            ⋮      ⋮     \n",
       "Cell7971 mydata_scRNAseq 36476      339          Batch4 Group1\n",
       "Cell7972 mydata_scRNAseq 35911      315          Batch4 Group1\n",
       "Cell7973 mydata_scRNAseq 21574      255          Batch4 Group1\n",
       "Cell7974 mydata_scRNAseq 12153      184          Batch4 Group1\n",
       "Cell7975 mydata_scRNAseq 23563      266          Batch4 Group1\n",
       "Cell7976 mydata_scRNAseq 26029      266          Batch4 Group1\n",
       "Cell7977 mydata_scRNAseq 34293      297          Batch4 Group2\n",
       "Cell7978 mydata_scRNAseq 24028      268          Batch4 Group1\n",
       "Cell7979 mydata_scRNAseq 19475      223          Batch4 Group1\n",
       "Cell7980 mydata_scRNAseq 20818      234          Batch4 Group1\n",
       "Cell7981 mydata_scRNAseq 18762      213          Batch4 Group1\n",
       "Cell7982 mydata_scRNAseq 36168      316          Batch4 Group2\n",
       "Cell7983 mydata_scRNAseq 31169      303          Batch4 Group1\n",
       "Cell7984 mydata_scRNAseq 38767      324          Batch4 Group2\n",
       "Cell7985 mydata_scRNAseq 40214      339          Batch4 Group1\n",
       "Cell7986 mydata_scRNAseq 25817      266          Batch4 Group1\n",
       "Cell7987 mydata_scRNAseq 22567      249          Batch4 Group1\n",
       "Cell7988 mydata_scRNAseq 28380      283          Batch4 Group1\n",
       "Cell7989 mydata_scRNAseq 32846      297          Batch4 Group1\n",
       "Cell7990 mydata_scRNAseq 30599      285          Batch4 Group1\n",
       "Cell7991 mydata_scRNAseq 57314      391          Batch4 Group1\n",
       "Cell7992 mydata_scRNAseq 26472      253          Batch4 Group2\n",
       "Cell7993 mydata_scRNAseq 44935      385          Batch4 Group1\n",
       "Cell7994 mydata_scRNAseq 39253      328          Batch4 Group2\n",
       "Cell7995 mydata_scRNAseq 27576      287          Batch4 Group1\n",
       "Cell7996 mydata_scRNAseq 16254      195          Batch4 Group1\n",
       "Cell7997 mydata_scRNAseq 42086      344          Batch4 Group1\n",
       "Cell7998 mydata_scRNAseq 37344      331          Batch4 Group1\n",
       "Cell7999 mydata_scRNAseq 25708      269          Batch4 Group1\n",
       "Cell8000 mydata_scRNAseq 21201      225          Batch4 Group2"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "mydata[[]]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "600a520b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# print(mydata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "d8a29c86",
   "metadata": {},
   "outputs": [],
   "source": [
    "# split the dataset into a list of batch object\n",
    "data.list <- SplitObject(mydata, split.by = \"Batch\")\n",
    "\n",
    "# normalize and identify variable features for each dataset independently\n",
    "data.list <- lapply(X = data.list, FUN = function(x) {\n",
    "    x <- NormalizeData(x)\n",
    "    x <- FindVariableFeatures(x, nfeatures = 5000)\n",
    "})\n",
    "\n",
    "# select features that are repeatedly variable across datasets for integration\n",
    "features <- SelectIntegrationFeatures(object.list = data.list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "b0accb34",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Scaling features for provided objects\n",
      "\n",
      "Finding all pairwise anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4752 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 753 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4740 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 730 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 7564 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 1169 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 4917 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 525 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 8822 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 938 anchors\n",
      "\n",
      "Running CCA\n",
      "\n",
      "Merging objects\n",
      "\n",
      "Finding neighborhoods\n",
      "\n",
      "Finding anchors\n",
      "\n",
      "\tFound 8804 anchors\n",
      "\n",
      "Filtering anchors\n",
      "\n",
      "\tRetained 983 anchors\n",
      "\n"
     ]
    }
   ],
   "source": [
    "immune.anchors <- FindIntegrationAnchors(object.list = data.list, anchor.features = features)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "37eff971",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Merging dataset 1 into 2\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 3 into 2 1\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n",
      "Merging dataset 4 into 2 1 3\n",
      "\n",
      "Extracting anchors for merged samples\n",
      "\n",
      "Finding integration vectors\n",
      "\n",
      "Finding integration vector weights\n",
      "\n",
      "Integrating data\n",
      "\n"
     ]
    }
   ],
   "source": [
    "all_features <- lapply(data.list, row.names) %>% Reduce(intersect, .) \n",
    "immune.combined <- IntegrateData( \n",
    "  anchorset = immune.anchors, \n",
    "  features.to.integrate = all_features,\n",
    "  verbose = T\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "f85bf9f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# immune.combined <- IntegrateData(anchorset = immune.anchors)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "59d4a662",
   "metadata": {},
   "outputs": [],
   "source": [
    "# specify that we will perform downstream analysis on the corrected data note that the\n",
    "# original unmodified data still resides in the 'RNA' assay\n",
    "# DefaultAssay(immune.combined) <- \"integrated\"\n",
    "\n",
    "# # Run the standard workflow for visualization and clustering\n",
    "# immune.combined <- ScaleData(immune.combined, verbose = FALSE)\n",
    "# immune.combined <- RunPCA(immune.combined, npcs = 30, verbose = FALSE)\n",
    "# immune.combined <- RunUMAP(immune.combined, reduction = \"pca\", dims = 1:30)\n",
    "# immune.combined <- FindNeighbors(immune.combined, reduction = \"pca\", dims = 1:30)\n",
    "# immune.combined <- FindClusters(immune.combined, resolution = 0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "e98561e6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Visualization\n",
    "# p1 <- DimPlot(immune.combined, reduction = \"umap\", group.by = \"batch\")\n",
    "# p2 <- DimPlot(immune.combined, reduction = \"umap\", label = TRUE, repel = TRUE)\n",
    "# p1 + p2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "e8e12c47",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assay data with 5000 features for 8000 cells\n",
       "Top 10 variable features:\n",
       " Gene1199, Gene4906, Gene2636, Gene4490, Gene3613, Gene1332, Gene3971,\n",
       "Gene2774, Gene3805, Gene320 "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "immune.combined$integrated"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "a947d5d4",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message:\n",
      "“Overwriting previous file seurat_simulation_1.loom”\n",
      "Adding col attribute CellID\n",
      "\n",
      "Adding col attribute orig.ident\n",
      "\n",
      "Adding col attribute nCount_RNA\n",
      "\n",
      "Adding col attribute nFeature_RNA\n",
      "\n",
      "Adding col attribute Batch\n",
      "\n",
      "Adding col attribute Group\n",
      "\n",
      "Adding row attribute Gene\n",
      "\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "Class: loom\n",
       "Filename: /disco_500t/rom/DisentangledContrastive/contrastive-sc/notebook/seurat_simulation_1.loom\n",
       "Access type: H5F_ACC_RDWR\n",
       "Listing:\n",
       "       name    obj_type dataset.dims dataset.type_class\n",
       "      attrs   H5I_GROUP         <NA>               <NA>\n",
       "  col_attrs   H5I_GROUP         <NA>               <NA>\n",
       " col_graphs   H5I_GROUP         <NA>               <NA>\n",
       "     layers   H5I_GROUP         <NA>               <NA>\n",
       "     matrix H5I_DATASET  8000 x 5000          H5T_FLOAT\n",
       "  row_attrs   H5I_GROUP         <NA>               <NA>\n",
       " row_graphs   H5I_GROUP         <NA>               <NA>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "pancreatic.loom <- as.loom(immune.combined, filename = \"seurat_simulation_1.loom\", verbose = FALSE, overwrite = TRUE)\n",
    "pancreatic.loom"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "0e672cca",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Always remember to close loom files when done\n",
    "pancreatic.loom$close_all()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "29113b22",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3df03f2f",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f824aeb0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
